backend/report/views.py

```diff
@@ -0,0 +1,299 @@
+                    start_time_list = answered_questionnaire_df.created_at.to_list()
+                    finish_time_list = finished_quetionnaires_df[finished_quetionnaires_df['Date'] == date].finished_at.to_list()
+                    time_diff_list = [(fn - st).seconds for fn, st in zip(finish_time_list, start_time_list)]
+                    questionaire_time_per_day = {"date": date, "total_time (seconds)": sum(time_diff_list)}
+                    period_questionnaire_time.append(questionaire_time_per_day)
+        if len(period_questionnaire_time) == 0:
+            return Response(data={"average_daily_questionnaire_time (seconds)": 0, "total_time (seconds)": []}, status=status.HTTP_200_OK)
+        else:
+            average_daily_questionnaire_time = sum([i['total_time (seconds)'] for i in period_questionnaire_time]) / len(period_questionnaire_time)
+            data = {"average_daily_questionnaire_time (seconds)": average_daily_questionnaire_time, "daily_questionnaire_time": period_questionnaire_time}
+            return Response(data=data, status=status.HTTP_200_OK)
+
+
+class AllUsersAverageAnnotationTimeAPI(APIView):
+    permission_classes = [IsAuthenticated & (IsProjectAdmin | IsProjectStaffAndReadOnly)]     
+
+    def get(self, request, *args, **kwargs):
+        all_user_ids = User.objects.all().values_list('id', flat=True)
+        startdate = self.kwargs["startdate"]
+        enddate = self.kwargs["enddate"]
+        progressed_users_avg_annotation_time = []
+        progressed_users_avg_annotation_time_per_day = []
+        for user_id in all_user_ids:
+            confirmed_examples_by_user = ExampleState.objects.get_confirmed_time_by_example([user_id], startdate, enddate)
+            if len(confirmed_examples_by_user) != 0:
+                confirmed_df = pd.DataFrame(confirmed_examples_by_user)
+                confirmed_df['Date'] = confirmed_df['confirmed_at'].dt.strftime('%Y-%m-%d')
+                confirmed_dates = list(confirmed_df.Date.unique())
+                daily_avg_single_annotation_time = []
+                all_single_annotation_time = []
+                for date in confirmed_dates:
+                    requested_examples = confirmed_df[confirmed_df['Date'] == date].example_id
+                    daily_single_annotation_time = []
+                    for example_id in requested_examples:
+                        started_examples_by_user = ExampleAnnotateStartState.objects.get_started_time_by_example(example_id, [user_id], startdate, enddate)
+                        annotation_time  = confirmed_df[confirmed_df['example_id']==example_id]["confirmed_at"] - started_examples_by_user[0]['started_at']
+                        daily_single_annotation_time.append(annotation_time.dt.seconds.values[0])
+                        all_single_annotation_time.append(annotation_time.dt.seconds.values[0])
```

Comment:
Somehow I got an index out of range error on `started_examples_by_user` when testing the APIs on the dump data from production. I managed to solve the issue on my local environment by putting line 194 - 196 under
```
if len(started_examples_by_user) > 0:
```

I think the error is due to there are some users not making any annotation at all? What do you think?

Reply:
I see. I get your point and I agree. I checked the dump data and this seems to be the culprit:
<img width="669" alt="bug" src="https://user-images.githubusercontent.com/64476430/211537521-a724b069-8855-410a-aa80-57cb7ab9b6a5.PNG">

User id 239 has confirmed example id 723, but there is no such record in table `examples_exampleannotatestartstate`. I also don't know why it can happen -- as you can see in the query result in the screenshot, this is the only one record where such a condition happens. Or maybe my dump data is corrupted somehow? XD please do check the dump data too, just in case 😅

Reply:
Yes it's not the problem of corrupted data, I have checked the data dump and for eg here, this user 84, on 10/12/22, he has confirmed examples from 221-241, but somehow there is no record of starting state for his examples 223, 232, 242 :sweat_smile: 
So because we need to see the report so I will adjust these endpoints a bit to handle those weird cases (but in such cases the statistics might be incorrect since we will discard some working time of them if there is no starting record even there should be) :sweat_smile:  
And we still need to investigate how can it happened :sweat_smile: 

Confirmed states
![Screenshot from 2023-01-10 14-20-47](https://user-images.githubusercontent.com/17596118/211566546-2d319f89-6899-4c35-b0bf-ff1b45548188.png) 

Started states
![Screenshot from 2023-01-10 14-20-37](https://user-images.githubusercontent.com/17596118/211566586-cc2d436b-68d2-4ae4-9cde-7578a9978458.png)